### PR TITLE
Backend/Utxo: support native segwit

### PR DESCRIPTION
**Pull request by aarani**

### CODE REVIEW

File: src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs [hunk 1]
```diff
@@ -197,17 +195,14 @@ module Account =
...
-                account.PublicKey.ScriptPubKey
-            else
-                //We filter utxos based on scriptPubKey when retrieving from electrum
-                //so this is unreachable.
-                failwith "Unreachable: unrecognized scriptPubKey"
-        )
-        :> ICoin
+        if sourceAddress = GetSegwitP2SHPublicAddressFromPublicKey account.Currency account.PublicKey then
+            coin.ToScriptCoin(account.PublicKey.WitHash.ScriptPubKey) :> ICoin
+        else if sourceAddress = GetNativeSegwitPublicAddressFromPublicKey account.Currency account.PublicKey then
```
Comment (knocte) [inline src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs:200]: @aarani elif please

File: src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs [hunk 2]
```diff
@@ -174,9 +190,24 @@ module Account =
...
+        // need functions that return scriptPubKey of our addresses instead of a
+        // string.
+        let sourceAddress = scriptPubKey.GetDestinationAddress(GetNetwork account.Currency).ToString()
         let coin =
             Coin(txHash, uint32 inputOutpointInfo.OutputIndex, Money(inputOutpointInfo.ValueInSatoshis), scriptPubKey)
-        coin.ToScriptCoin account.PublicKey.WitHash.ScriptPubKey :> ICoin
+        coin.ToScriptCoin(
+            if sourceAddress = GetSegwitP2SHPublicAddressFromPublicKey account.Currency account.PublicKey then
+                account.PublicKey.WitHash.ScriptPubKey
+            else if sourceAddress = GetNativeSegwitPublicAddressFromPublicKey account.Currency account.PublicKey then
```
Comment (knocte) [inline src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs]: @aarani elif please

File: src/GWallet.Backend/UtxoCoin/ElectrumClient.fs [hunk 3]
```diff
@@ -71,6 +71,49 @@ module ElectrumClient =
...
+        //        - query details of each element in Z to see their block heights
+        //        - query the current blockheight (H) -> pick the highest among all servers queried
+        //        -> having H, we now know which elements of X, Y, and Z are confirmed or not
+        // Doing it this way has two advantages:
+        // 1) We can configure GWallet with a number of confirmations to consider some balance confirmed (instead
+        //    of trusting what "confirmed" means from the point of view of the Electrum Server)
+        // 2) and most importantly: we could verify each of the transactions supplied in X, Y, Z to verify their
+        //    integrity (in a similar fashion as Electrum Wallet client already does), to not have to trust servers*
+        //    [ see https://www.youtube.com/watch?v=hjYCXOyDy7Y&feature=youtu.be&t=1171 for more information ]
+        // * -> although that would be fixing only half of the problem, we also need proof of completeness
```
Comment (knocte) [inline src/GWallet.Backend/UtxoCoin/ElectrumClient.fs:88]: @aarani this comment looks very familiar to me, like if I had written it, not you? are you copy+pasting it from somewhere? (and thus not respecting DRY?)

Comment (aarani) [reply src/GWallet.Backend/UtxoCoin/ElectrumClient.fs:88]: Yes it yours, look just above the hunk

Comment (aarani) [reply src/GWallet.Backend/UtxoCoin/ElectrumClient.fs:88]: see the code as far as I can tell I can't reuse anything from the other one

Comment (aarani) [reply src/GWallet.Backend/UtxoCoin/ElectrumClient.fs:88]: > why dont you delete the previous one then?

Good point.

Comment (aarani) [reply src/GWallet.Backend/UtxoCoin/ElectrumClient.fs:88]: I kinda wanted to do the least amount of damage until I got the gtg.

File: src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs [hunk 4]
```diff
@@ -292,9 +318,28 @@ module Account =
...
+            async {
+                let! nativeSegwitUtxos =
+                    GetNativeSegwitPublicAddressFromPublicKey currency account.PublicKey
+                    |> getUtxos
+
+                let! legacySegwitUtxos =
+                    GetSegwitP2SHPublicAddressFromPublicKey currency account.PublicKey
+                    |> getUtxos
+
+                return Array.concat [ nativeSegwitUtxos; legacySegwitUtxos ]
```
Comment (knocte) [inline src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs:341]: @aarani can use Seq. here?

File: src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs [hunk 5]
```diff
@@ -137,11 +145,17 @@ module Account =
                             (mode: ServerSelectionMode)
                             (cancelSourceOption: Option<CustomCancelSource>)
                                 : Async<BlockchainScriptHashGetBalanceInnerResult> =
-        let scriptHashHex = GetElectrumScriptHashFromPublicAddress account.Currency account.PublicAddress
+        let scriptHashesHex =
+            [
+                GetNativeSegwitPublicAddressFromPublicKey account.Currency account.PublicKey
+                    |> GetElectrumScriptHashFromPublicAddress account.Currency
+                GetSegwitP2SHPublicAddressFromPublicKey account.Currency account.PublicKey
```
Comment (knocte) [inline src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs:152]: nit, rename to GetSegwitP2shPublicAddressFromPublicKey

**Comment by knocte** [comment]
One last thing, can we do this in a soft-launch fashion? put a setting in Config.fs that is for now as =false, so geewallet behaves in same way as before. Then whenever we set it to =true, this new behaviour starts happening (so maybe we set it as true in lightning branch).

File: src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs (hunk 5 above)
Comment (aarani) [reply src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs:152]: sh is an acronym tho

Comment (aarani) [reply src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs:152]: it's technically Pay 2 Script Hash so I think it should be uppercase

Comment (knocte) [reply src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs:152]: there are many acronyms that, when used in .NET APIs, they still get PascalCasedLikeThis, and you tend to forget this fact

File: src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs (hunk 4 above)
Comment (aarani) [reply src/GWallet.Backend/UtxoCoin/UtxoCoinAccount.fs]: Done

**Comment by aarani** [comment]
> One last thing, can we do this in a soft-launch fashion? put a setting in Config.fs that is for now as =false, so geewallet behaves in same way as before. Then whenever we set it to =true, this new behaviour starts happening (so maybe we set it as true in lightning branch).

Done

File: src/GWallet.Backend/Config.fs [hunk 6]
```diff
@@ -35,6 +35,8 @@ module Config =
     //       balances, so you might find discrepancies (e.g. the donut-chart-view)
     let internal NoNetworkBalanceForDebuggingPurposes = false
 
+    let internal UseNativeSegwit = false
```
Comment (knocte) [inline src/GWallet.Backend/Config.fs:38]: @aarani given we're still leaving this as false by default, please update commit message to not be misleading

**Comment by knocte** [comment]
Title's scope should be Backend/Utxo:, not just Backend.

**Comment by knocte** [comment]
CI is red

**Comment by knocte** [comment]
@aarani after you finish with the `--console` task, can you please add a commit to this PR that does the following:

1. Make UseNativeSegwit default to true if you pass certain parameter to ./configure.sh (e.g. --native-segwit), which in turn would make msbuild/xbuild/dotnet build with DefineConstants+=NATIVE_SEGWIT .
2. Make CI have additional steps that build with this constant turned on, and publish the result in the beta channel (as opposed to the stable channel). See https://github.com/nblockchain/geewallet/commit/37f9c20cd4311be3ebec9e07950448d4bd556689 for context.

**Comment by knocte** [comment]
> Make CI have additional steps that build with this constant turned on...

Just to make sure you understand the above, take in account that I want the build to publish 2 snaps: one for the stable channel and one for the beta one.

File: scripts/make.fsx [hunk 7]
```diff
@@ -136,7 +136,7 @@ let GetOrExplain key map =
 let prefix = buildConfigContents |> GetOrExplain "Prefix"
 let libPrefixDir = DirectoryInfo (Path.Combine (prefix, "lib", UNIX_NAME))
 let binPrefixDir = DirectoryInfo (Path.Combine (prefix, "bin"))
-
+
```
Comment (knocte) [inline scripts/make.fsx:139]: @aarani unneeded

**Comment by knocte** [comment]
I see conflicts. Also please do not duplicate a whole script, you can pass the --native-segwit arg to the script, and then that script can pass all args that it receives to the ./configure call.

**Comment by knocte** [comment]
If CI becomes green here, I guess the last things to do are (in strict order):

1. Make beta snap have different version than stable snap (e.g. if stable is 0.5.x.y, then 0.7.x.y)
2. Test that everything works with a tag push in your fork.
3. Remove WIP from commit msg.

File: .github/workflows/CI.yml [hunk 8]
```diff
@@ -576,6 +576,7 @@ jobs:
     - name: Upload snap package to Snap Store
       env:
         SNAPCRAFT_LOGIN: ${{ secrets.SNAPCRAFT_LOGIN }}
+        OVERRIDE_SNAP_VERSION: 0.7.2.1
```
Comment (knocte) [inline .github/workflows/CI.yml:579]: @aarani harcoding a version number? are you serious????

Comment (knocte) [reply .github/workflows/CI.yml:579]: @Mersho ^

File: scripts/snap_release.fsx [hunk 9]
```diff
@@ -159,6 +159,19 @@ let snapPush =
...
             failwith "Invalid arguments"
 
+    let overrideVersion = Environment.GetEnvironmentVariable "OVERRIDE_SNAP_VERSION"
+    if not (String.IsNullOrWhiteSpace overrideVersion) then
+        let snapcraftDir = Path.Combine(FsxHelper.RootDir.FullName, "snap", "snapcraft.yaml")
+        let readSnapcraftYaml = File.ReadAllText snapcraftDir
+        let matchVersion = Regex.Match(readSnapcraftYaml, @"(?<=version:\s')[\d\.]+(?=')")
+        if matchVersion.Success then
+            let versionList = matchVersion.Value.Split('.')
+            Array.set versionList 1 ((int versionList.[1] + 2).ToString())
```
Comment (aarani) [inline scripts/snap_release.fsx:169]: I hate this line

Comment (knocte) [reply scripts/snap_release.fsx:169]: @aarani please elaborate ; @Mershad: first question, does it work? second: if yes, please add a comment when you have to write such an unreadable code, man. third: I'll further review later when I get replies from the above and my comments are addressed

Comment (Mersho) [reply scripts/snap_release.fsx:169]: This code reads snapcraft.yaml and rewrites the "version".
Previously, we used `snapcraftctl set-version` to set the version, but I'm not sure it works.
@knocte

Comment (knocte) [reply scripts/snap_release.fsx:169]: stop using thinking about using that weird command please; setting the version in the YAML has always worked, so using a different method here doesn't make any sense; btw you haven't answered yet my first question btw

Comment (Mersho) [reply scripts/snap_release.fsx:169]: I tested the code on Interactive console, but it does not change on CI.

**Comment by knocte** [comment]
Any why the scope of the last commit is completely wrong??????

**Pull request merged**